sotoki/sotoki.py

```diff
@@ -879,30 +889,64 @@ def get_filetype(headers, path):
             ftype = "ico"
     return ftype
 
+def download_from_cache(key, output, size, cache_storage):
+    ret = False
+    if cache_storage.has_object_matching_meta(key, "size", str(size)):
+        try:
+            cache_storage.download_file(key, output, progress=True)
+            print("Successfully downloaded already optimized version from cache.")
+            #cache_storage.set_object_autodelete_on(key, datetime.datetime.now() + datetime.timedelta(days=30))
+            #Currently gives 403 forbidden
+            ret = True
+        except Exception as e:
+            print("Failed to download from cache\n" + str(e))
+            ret = False
+    else:
+        print("Optimized file not found in cache")
+        ret = False
+    return ret
 
-def download_image(url, fullpath, convert_png=False, resize=False):
-    headers = None
-    tmp_img = None
+def upload_to_cache(fpath, key, size, cache_storage):
     try:
-        tmp_img = get_tempfile(os.path.basename(fullpath))
-        headers = download(url, tmp_img, timeout=60)
-    except urllib.error.URLError as e:
-        os.unlink(tmp_img)
-        print("Cannot download " + fullpath)
-        print(e)
-    else:
-        ext = get_filetype(headers, tmp_img)
+        cache_storage.upload_file(fpath, key, meta={"size" : str(size)})
+        #cache_storage.set_object_autodelete_on(key, datetime.datetime.now() + datetime.timedelta(days=30))
+        #Currently gives 403 forbidden
+    except Exception as e:
+        raise Exception("Failed to upload to cache\n" + str(e))
+    
+
+def download_image(url, fullpath, convert_png=False, resize=False, cache_storage=None):
+    downloaded = False
+    key = None
+    if cache_storage is not None:
+        plain_url = re.sub("^https?://", "", url)
+        key = urllib.parse.quote(plain_url)
```

Comment:
Will change this to {domain name that we are parsing}/{path} (the path will be relative to the output folder)
@rgaudin do you have something better in mind? I'm open to suggestions.